### PR TITLE
Support building 'Dockerfile' of container actions on k8s runner using Kaniko

### LINKED ISSUE #23: Support building container actions in k8s hooks
Currently, we don't support building container actions for k8s hooks. Is there a way we could support it? Maybe via a tool like https://github.com/GoogleContainerTools/kaniko and push them up to the container registry.

### PR DESCRIPTION
**Pull request by fhammerl**
@nikola-jokic 
@fhammerl

Using [kaniko](https://github.com/GoogleContainerTools/kaniko), we can build a Dockerfile in k8s.

This unlocks container actions that build their own images (instead of referencing one in a registry).

https://github.com/fhammerl/noff/actions/runs/3128336666/jobs/5076100801 👇
image

### Mini ADR

WIP

Closes #23

### CODE REVIEW

File: packages/k8s/src/k8s/index.ts [hunk 1]
```diff
@@ -464,6 +468,16 @@ export async function isPodContainerAlpine(
   return isAlpine
 }
 
+export async function containerBuild(
+  args: RunContainerStepArgs,
+  imagePath: string
+): Promise<string> {
+  const registryUri = `localhost:${registryNodePort()}/${imagePath}`
```
Comment (fhammerl) [inline packages/k8s/src/k8s/index.ts:475]: Consider allowing external repositories as well:
1. `localhost:${registryNodePort()}/${imagePath}` = `localhost:31500/handl/image:aiwefjliw389`

2. `ghcr.io/fhammerl/image:aiwefjliw389`

File: packages/k8s/src/k8s/index.ts [hunk 2]
```diff
@@ -326,7 +323,14 @@ export async function waitForPodPhases(
   let phase: PodPhase = PodPhase.UNKNOWN
   try {
     while (true) {
-      phase = await getPodPhase(podName)
+      try {
+        phase = await getPodPhase(podName)
+      } catch (err) {
+        const e = err as k8s.HttpError
```
Comment (nikola-jokic) [inline packages/k8s/src/k8s/index.ts:329]: This is the case when we are creating for example a `StatefulSet` for `docker-registry`. The controller is about to create a pod, but we can hit this function potentially before the reconciliation happens. I don't think that we should handle it this way, I just made it like this until we figure out how to proceed.

One way is to add retries in this case. The other solution is to have pod listing of the outer resource, so we wait there before we actually await for the pod phase. Not sure which one should be more appropriate. Maybe there is a better way that I am not seeing right now :relaxed:

File: packages/k8s/src/hooks/run-container-step.ts [hunk 3]
```diff
@@ -108,3 +111,20 @@ function createPodSpec(
...
+
+function generateBuildTag(): string {
+  return `${generateRandomString()}:${generateRandomString(6)}`
+}
+
+function generateBuildHandle(): string {
+  return generateRandomString()
+}
+
+function generateRandomString(length = 10): string {
```
Comment (TingluoHuang) [inline packages/k8s/src/hooks/run-container-step.ts:123]: any library we can leverage for generating this? what about using UUID?

File: packages/k8s/src/k8s/settings.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+export const POD_VOLUME_NAME = 'work'
+export function namespace(): string {
+  if (process.env['ACTIONS_RUNNER_KUBERNETES_NAMESPACE']) {
+    return process.env['ACTIONS_RUNNER_KUBERNETES_NAMESPACE']
+  }
+
+  const context = kc.getContexts().find(ctx => ctx.namespace)
+  if (!context?.namespace) {
+    throw new Error(
+      'Failed to determine namespace, falling back to `default`. Namespace should be set in context, or in env variable "ACTIONS_RUNNER_KUBERNETES_NAMESPACE"'
```
Comment (TingluoHuang) [inline packages/k8s/src/k8s/settings.ts:19]: are we falling back to default or throwing an error and failing the step?

File: packages/k8s/src/hooks/run-container-step.ts (hunk 3 above)
Comment (nikola-jokic) [reply packages/k8s/src/hooks/run-container-step.ts]: Applied :relaxed:

**Pull request closed without merging**